src/wasm/wasm-type.cpp

```diff
@@ -1180,6 +1209,220 @@ bool SubTyper::isSubType(const Rtt& a, const Rtt& b) {
   return a.heapType == b.heapType && a.hasDepth() && !b.hasDepth();
 }
 
+bool TypeBounder::hasLeastUpperBound(Type a, Type b) {
+  Type tempLUB;
+  return lub(a, b, tempLUB);
+}
+
+Type TypeBounder::getLeastUpperBound(Type a, Type b) {
+  Type tempLUB;
+  if (!lub(a, b, tempLUB)) {
+    return Type::none;
+  }
+  // `tempLUB` is a temporary type owned by `builder`. Since TypeBuilder::build
+  // returns HeapTypes rather than Types, create a new HeapType definition meant
+  // only to get `tempLUB` canonicalized in a known location.
+  builder.grow(1);
+  builder[builder.size() - 1] = Array(Field(tempLUB, Mutable));
+  return builder.build().back().getArray().element.type;
+}
+
+bool TypeBounder::lub(Type a, Type b, Type& out) {
+  if (a == b) {
+    out = a;
+    return true;
+  }
+  if (a == Type::unreachable) {
+    out = b;
+    return true;
+  }
+  if (b == Type::unreachable) {
+    out = a;
+    return true;
+  }
+  if (a.isTuple() && b.isTuple()) {
+    Tuple tuple;
+    if (!lub(a.getTuple(), b.getTuple(), tuple)) {
+      return false;
+    }
+    out = builder.getTempTupleType(tuple);
```

Comment:
why is it ok to return temp types in all these places?

Reply:
`lub` methods are only called by other `lub` methods, and by `TypeBounder::{get,has}LeastUpperBound`, none of which allow the temporary types to escape outside of the `TypeBounder` and its `TypeBuilder`.

Reply:
I'll make the `lub` methods private to enforce that more.